### ISSUE TITLE
Add support for buildpacks API 0.10

### ISSUE THREAD

**Issue by dmikusa** [buildpacks/libcnb#260]
This could probably be broken down into more tasks, but at a high level we need to add support for 0.10.

https://buildpacks.io/docs/reference/spec/migration/buildpack-api-0.9-0.10/

**Comment by loewenstein** [comment]
Doesn't it need to be about 0.8 -> 0.9 -> 0.10, i.e. using https://buildpacks.io/docs/reference/spec/migration/buildpack-api-0.8-0.9/ and then https://buildpacks.io/docs/reference/spec/migration/buildpack-api-0.9-0.10/?

**Comment by nicolasbender** [comment]
> This could probably be broken down into more tasks

Could you provide more details for these subtasks, please? We haven't found any relevant changes from 0.9 -> 0.10 affecting libcnb.
Out of interest: Why was `v2` of `libcnb` created in the first place?

**Comment by dmikusa** [comment]
@loewenstein 0.9 is already supported, so we just need to add 0.10 support. It needs to be an additive change.

@nicolasbender 

- [ ] We need to mark anything related to stacks as deprecated. I haven't looked to see exactly what's in there, but there is [some stuff](https://github.com/search?q=repo%3Abuildpacks%2Flibcnb%20stack&type=code).
- [ ] We need to add support for targets. This requires [toml, struct, and serialization/deserialization changes](https://buildpacks.io/docs/for-buildpack-authors/how-to/migrate/buildpack-api-0.9-0.10/#before-build). 
- [ ] There are new target-related env variables exposed. With [stacks we put that into the context](https://github.com/buildpacks/libcnb/blob/main/build.go#L228-L232), we should probably do something similar with targets.
- [ ] The migration guide doesn't [elaborate on runtime image extensions](https://buildpacks.io/docs/for-buildpack-authors/how-to/migrate/buildpack-api-0.9-0.10/#run-image-extensions-is-supported-experimental), but this needs to be spiked to see if there are any required changes to support this. For example, there might be struct/toml changes.

> Out of interest: Why was v2 of libcnb created in the first place?

In a nutshell, breaking API changes. We reworked the core build/detect mechanism. It was previously done with structs at its heart, and we changed it to be function-based now. I believe it'll be more flexible, but this was also done in conjunction with the Paketo team to position them to be able to use libcnb as the base for their packit library.

Aside from that, we've removed some other cruft like old buildpack API compatibility.

**Comment by loewenstein** [comment]
> * The migration guide doesn't [elaborate on runtime image extensions](https://buildpacks.io/docs/for-buildpack-authors/how-to/migrate/buildpack-api-0.9-0.10/#run-image-extensions-is-supported-experimental), but this needs to be spiked to see if there are any required changes to support this. For example, there might be struct/toml changes.

Is libcnb supposed to support authoring extensions?

**Comment by c0d1ngm0nk3y** [comment]
> It was previously done

But why wasn't `v2` then released earlier? `v1` doesn't support the Buildpack API 0.10 either, does it?

**Comment by dmikusa** [comment]
v1 supports 0.8. That is the highest it will ever go because going above that requires breaking API changes.

https://github.com/buildpacks/libcnb/blob/release-1.x/build.go#L95-L99

Yes, we support extensions in v2. There are a couple under Paketo Community using it, but we haven't had a huge demand for it. I was hoping that it would stabilize more before we release libcnb v2 and it looks like it has so that's another milestone complete towards v2 release.
